src/bms/parse/random/ast_build.rs

```diff
@@ -0,0 +1,591 @@
+            Token::Random(_) | Token::SetRandom(_) => {
+                result.push(parse_random_block(iter, error_list));
+            }
+            Token::EndIf => {
+                error_list.push(ControlFlowRule::UnmatchedEndIf);
+                iter.next();
+            }
+            Token::EndRandom => {
+                error_list.push(ControlFlowRule::UnmatchedEndRandom);
+                iter.next();
+            }
+            _ => {
+                // Directly collect non-control-flow tokens
+                if !token.is_control_flow_token() {
+                    result.push(Unit::Token(*token));
+                }
+                iter.next();
+            }
+        }
+    }
+    result
+}
+
+/// Parse a Random/SetRandom block until EndRandom
+fn parse_random_block<'a, I>(
+    iter: &mut std::iter::Peekable<I>,
+    error_list: &mut Vec<ControlFlowRule>,
+) -> Unit<'a>
+where
+    I: Iterator<Item = &'a Token<'a>>,
+{
+    let token = iter.next().unwrap();
+    let block_value = match token {
+        Token::Random(val) => BlockValue::Random { max: *val as u64 },
+        Token::SetRandom(val) => BlockValue::Set { value: *val as u64 },
+        _ => unreachable!(),
+    };
+    let mut if_blocks = Vec::new();
+    while let Some(next) = iter.peek() {
+        match next {
+            Token::If(if_val) => {
+                iter.next();
+                let (tokens, _has_endif) = parse_if_block_body(iter, error_list);
+                let mut branches = HashMap::new();
+                branches.insert(
+                    *if_val as u64,
+                    IfBranch {
+                        value: *if_val as u64,
+                        tokens,
+                    },
+                );
+                // Parse ElseIf branches
+                while let Some(Token::ElseIf(elif_val)) = iter.peek() {
+                    iter.next();
+                    let (elif_tokens, _has_endif) = parse_if_block_body(iter, error_list);
+                    branches.insert(
+                        *elif_val as u64,
+                        IfBranch {
+                            value: *elif_val as u64,
+                            tokens: elif_tokens,
+                        },
+                    );
+                }
+                // Check for unmatched ElseIf
+                if let Some(Token::ElseIf(_)) = iter.peek() {
+                    error_list.push(ControlFlowRule::UnmatchedElseIf);
+                    iter.next();
+                }
+                // Parse Else branch
+                if let Some(Token::Else) = iter.peek() {
+                    iter.next();
+                    let (etokens, _has_endif) = parse_if_block_body(iter, error_list);
+                    branches.insert(
+                        0,
+                        IfBranch {
+                            value: 0,
+                            tokens: etokens,
+                        },
+                    );
+                }
```

Comment:
On lexicalize BMS source, the value 0 is allowed for `#IF` and `#ELSEIF`. To prevent occurring ambiguity, value 0 in `#IF` and `#ELSEIF` should be treated as an error.